feat: group main help message to put important commands to the top

grafvonb opened the pull request:
```release-notes
Groups root help message to include important commands at the top
```

lance commented:
@grafvonb how is it going with this? When you are ready for a review, please remove the WIP label and title description. Thanks!

grafvonb commented:
> @grafvonb how is it going with this? When you are ready for a review, please remove the WIP label and title description. Thanks!

Hi, I'm still working on so the WIP label is right. I hope I will get it ready in the following days and than I will change the label as requested.

lance commented on .gitignore:
```diff
@@ -1,18 +1,31 @@
+# Non relevant vendor files
+BUILD.bazel
+OWNERS
+OWNERS_ALIASES
```
Although these additions are useful, I would prefer this to be in a separate PR as they are unrelated to the feature you are adding. Not a deal breaker, but that is my preference.

grafvonb replied:
as requested: https://github.com/knative-sandbox/kn-plugin-func/pull/1209

lance reviewed (commented):
Overall this looks good. Thanks for the work on this! I have a few small suggestions in the code, and ask that you also run `make docs` since this will possibly affect the generated documentation.

lance commented on cmd/templates/command_groups_test.go:
```diff
@@ -0,0 +1,69 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templates
```
Thanks for moving this to `cmd/templates`! I did not really think about the `templates` location in my previous review, but having these `.go` files in there was not right. This is much better. Thanks!

grafvonb commented:
@lance Hi, I found some time today to continue working on the feature. As you know for the implementation I use the code from `kn` as a basis and have discovered two problems today, so the ticket is back in WIP status.

The first problem is rather small, it's about the wording. `kn` uses the word 'options' as a replacement for 'flags'. Should we also use 'options' for `func`? 

The second is about listing help for the global flags (options) like '--help', '--namespace' etc. 
`kn` uses a command `kn options` for this ("The following options can be passed to any command:"), otherwise only a static hint is displayed. How do we deal with this?

grafvonb commented:
@lance Maybe to keep it simple I will implement exactly the same help behaviour as current one (cobra's help standard), means: 
1. changing all current wording "inherited" form the `kn` implementation from `options` to `flags`,
2. expand help templates code to be able to list global flags as current implementation does.
What do you think?
PS. As I written above, an alternative for point 2 would be a new command that lists the global flags on demand (aka `kn` style)

lance commented:
> @lance Maybe to keep it simple I will implement exactly the same help behaviour as current one (cobra's help standard), means:
> 
> 1. changing all current wording "inherited" form the `kn` implementation from `options` to `flags`,

I think this is probably the best course of action for now.

> 2. expand help templates code to be able to list global flags as current implementation does.

This should be pretty straightforward, right?

>    What do you think?
>    PS. As I written above, an alternative for point 2 would be a new command that lists the global flags on demand (aka `kn` style)

I think it's odd, I suppose. I am not sure how intuitive `func options` or `func flags` would be for a user vs. simply `func <cmd> --help`. Frankly, I've been using `kn` for a while now and did not know about the `options` command. In any case, we really only have `--help` and `--verbose`, I think. Perhaps `--path`.

My preference is to expand the templates code as you suggest.

grafvonb commented:
@lance The feature is actually ready, as discussed I changed the labels to `flags` and replaced options command by global flags template. However after docs generation I found two another problems:
1. in the generated Synopsis section for all commands there is no `[flags]` suffix. I will have to change the implementation to achieve that,
2. it is an observation, the generated docs uses `options` instead of `flags`. I suppose that was the reason why `kn` help implementation changes the wording. We have no possibility to change this behaviour as it is hard coded in `docs` package of Cobra:

```
func printOptions(buf *bytes.Buffer, cmd *cobra.Command, name string) error {
	flags := cmd.NonInheritedFlags()
	flags.SetOutput(buf)
	if flags.HasAvailableFlags() {
		buf.WriteString("### Options\n\n```\n")
		flags.PrintDefaults()
		buf.WriteString("```\n\n")
	}

	parentFlags := cmd.InheritedFlags()
	parentFlags.SetOutput(buf)
	if parentFlags.HasAvailableFlags() {
		buf.WriteString("### Options inherited from parent commands\n\n```\n")
		parentFlags.PrintDefaults()
		buf.WriteString("```\n\n")
	}
	return nil
}
```
It would be good to have a consistent wording in the `func` help and docs, however it is not critical to me.

lance commented:
@grafvonb I see what you mean. I agree that consistent wording is important, but I don't want to hold up this PR on that issue. The other issue you note

>in the generated Synopsis section for all commands there is no [flags] suffix. I will have to change the implementation to achieve that

seems also to be a minor thing to me. It's the difference between the string `func build` and `func build [flags]` as even in the existing generated documentation, the flags themselves are not enumerated here but rather in the Options section. I think it's OK to go with both of these as is for now, and open two new issues to address them. @zroubalik @matejvasek @lkingland @salaboy wdyt?

zroubalik commented:
Agree 👍

lance commented:
/lgtm

lance commented:
@grafvonb are you ready for this to merge?

grafvonb commented:
@lance I just fixed a problem where the docs for the root command was not generated correctly. The problem was that we were using placeholders in the "long" help message, which was not supported by the `kn` version. I also moved the help messages back to the main command, because storing them in the templates breaks the coherence. Now the help is displayed correctly and the docs are fine too. I think I am done with the problem.

zroubalik reviewed (approved):
/lgtm
/approve

/hold for others

Thanks a lot for this contribution!

lance commented:
/lgtm

Thanks @grafvonb for the contribution!

lance commented:
/unhold
